File: common/src/main/java/generations/gg/generations/core/generationscore/world/level/block/utilityblocks/DyeableBlock.java
package generations.gg.generations.core.generationscore.world.level.block.utilityblocks;

import dev.architectury.registry.registries.RegistrySupplier;
import generations.gg.generations.core.generationscore.world.level.block.entities.ModelProvidingBlockEntity;
import generations.gg.generations.core.generationscore.world.level.block.entities.MutableBlockEntityType;
import generations.gg.generations.core.generationscore.world.level.block.generic.GenericRotatableModelBlock;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.DyeColor;
import net.minecraft.world.item.DyeItem;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.storage.loot.LootParams;
import net.minecraft.world.phys.BlockHitResult;
import org.jetbrains.annotations.NotNull;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;

@SuppressWarnings("deprecation")
public abstract class DyeableBlock<T extends ModelProvidingBlockEntity, V extends DyeableBlock<T, V>> extends GenericRotatableModelBlock<T> {
    private final DyeColor color;
    private final Map<DyeColor, RegistrySupplier<V>> function;

    public DyeableBlock(DyeColor color, Map<DyeColor, RegistrySupplier<V>> function, RegistrySupplier<MutableBlockEntityType<T>> biFunction, BiFunction<BlockPos, BlockState, BlockPos> baseBlockPosFunction, Properties arg, ResourceLocation model, int width, int height, int length) {
        super(arg, biFunction, baseBlockPosFunction, model, width, height, length);
        this.color = color;
        this.function = function;
    }

    public DyeableBlock(DyeColor color, Map<DyeColor, RegistrySupplier<V>> function, RegistrySupplier<MutableBlockEntityType<T>> biFunction, BiFunction<BlockPos, BlockState, BlockPos> baseBlockPosFunction, Properties arg, ResourceLocation model) {
        super(arg, biFunction, baseBlockPosFunction, model);
        this.color = color;
        this.function = function;
    }

    public DyeableBlock(DyeColor color, Map<DyeColor, RegistrySupplier<V>> function, RegistrySupplier<MutableBlockEntityType<T>> biFunction, Properties arg, ResourceLocation model, int width, int height, int length) {
        super(arg, biFunction, model, width, height, length);
        this.color = color;
        this.function = function;
    }

    public DyeableBlock(DyeColor color, Map<DyeColor, RegistrySupplier<V>> function, RegistrySupplier<MutableBlockEntityType<T>> biFunction, Properties arg, ResourceLocation model) {
        super(arg, biFunction, model);
        this.color = color;
        this.function = function;
    }

    @Override
    public @NotNull InteractionResult use(@NotNull BlockState state, Level world, @NotNull BlockPos pos, @NotNull Player player, @NotNull InteractionHand handIn, @NotNull BlockHitResult hit) {
        if (!world.isClientSide() && handIn == InteractionHand.MAIN_HAND) {
            if (!tryDyeColor(state, world, pos, player, handIn, hit)) {

                return serverUse(state, (ServerLevel) world, pos, (ServerPlayer) player, handIn, hit);
            }
            else return InteractionResult.SUCCESS;
        }
        return InteractionResult.FAIL;
    }

    public Item getItemFromDyeColor(DyeColor color) {
        return getBlockFromDyeColor(color).asItem();
    }

    public V getBlockFromDyeColor(DyeColor color) {
        return function.get(color).get();
    }

    public boolean tryDyeColor(@NotNull BlockState state, Level world, @NotNull BlockPos pos, @NotNull Player player, @NotNull InteractionHand handIn, @NotNull BlockHitResult hit) {
        ItemStack heldItem = player.getMainHandItem();

        boolean isEmpty = !heldItem.isEmpty();
        boolean isDye = heldItem.getItem() instanceof DyeItem;

        if (isEmpty && isDye) {

            DyeColor dyeColor = ((DyeItem) heldItem.getItem()).getDyeColor();

            var baseState = world.getBlockState(pos);

            if(this.getClass().isInstance(baseState.getBlock())) {
                var baseBlock = this.getClass().cast(baseState.getBlock());


                if (!baseBlock.color.equals(dyeColor)) {

                    var base = getBaseBlockPos(pos, state);

                    if (!player.isCreative()) heldItem.shrink(1);

                    var newBlock = getBlockFromDyeColor(dyeColor);

                    var defaultState = newBlock.defaultBlockState().setValue(FACING, baseState.getValue(FACING));

                        var dir = state.getValue(FACING);

                        for (int x = 0; x < width + 1; x++) {
                            for (int y = 0; y < height + 1; y++) {
                                for (int z = 0; z < length + 1; z++) {
                                    var adjustX = adjustX(x);
                                    var adjustZ = adjustX(x);

                                    var blockPos = base.relative(dir.getCounterClockWise(), adjustX).relative(Direction.UP, y).relative(dir, adjustZ);

                                    var currentState = world.getBlockState(blockPos);

                                    var stateX = baseBlock.getWidthValue(currentState);
                                    var stateY = baseBlock.getHeightValue(currentState);
                                    var stateZ = baseBlock.getLengthValue(currentState);

                                    world.setBlock(blockPos, newBlock.setSize(defaultState.setValue(WATERLOGGED, currentState.getValue(WATERLOGGED)), stateX, stateY, stateZ), 2, 0);
                                }
                            }
                        }
                        return true;
                }
            }
        }

        return false;
    }

    @Override
    public @NotNull ItemStack getCloneItemStack(BlockGetter blockGetter, BlockPos blockPos, BlockState blockState) {
        return new ItemStack(getItemFromDyeColor(getColor()));
    }

    protected InteractionResult serverUse(BlockState state, ServerLevel world, BlockPos pos, ServerPlayer player, InteractionHand handIn, BlockHitResult hit) {
        return InteractionResult.PASS;
    }

    public DyeColor getColor() {
        return color;
    }

    @Override
    public @NotNull List<ItemStack> getDrops(@NotNull BlockState state, @NotNull LootParams.Builder params) {
        return getWidthValue(state) == 0 || getHeightValue(state) == 0 || getLengthValue(state) == 0 ? super.getDrops(state, params) : Collections.emptyList();
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/level/block/utilityblocks/DyeableBlock.kt
package generations.gg.generations.core.generationscore.common.world.level.block.utilityblocks

import com.mojang.datafixers.Products
import com.mojang.serialization.Codec
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.instanceOrNull
import generations.gg.generations.core.generationscore.common.world.level.block.entities.ModelProvidingBlockEntity
import generations.gg.generations.core.generationscore.common.world.level.block.generic.GenericRotatableModelBlock
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.core.Holder
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.InteractionHand
import net.minecraft.world.ItemInteractionResult
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.DyeColor
import net.minecraft.world.item.DyeItem
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level
import net.minecraft.world.level.LevelReader
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.storage.loot.LootParams
import net.minecraft.world.phys.BlockHitResult

fun <T: Any> T.applyIfTrue(predicate: (T) -> Boolean, action: (T) -> T): T {
    if(predicate.invoke(this)) return action.invoke(this)
    return this
}

abstract class DyeableBlock<T : ModelProvidingBlockEntity, V : DyeableBlock<T, V>> : GenericRotatableModelBlock<T> {
    @JvmField
    val color: DyeColor
    val map: Map<DyeColor, Block>

    constructor(
        color: DyeColor,
        function: Map<DyeColor, Block>,
        baseBlockPosFunction: (BlockPos, BlockState) -> BlockPos,
        arg: Properties,
        model: ResourceLocation,
        width: Int,
        height: Int,
        length: Int
    ) : super(arg, baseBlockPosFunction, model, width, height, length) {
        this.color = color
        this.map = function
    }

    constructor(
        color: DyeColor,
        function: Map<DyeColor, Block>,
        baseBlockPosFunction: (BlockPos, BlockState) -> BlockPos,
        arg: Properties,
        model: ResourceLocation
    ) : super(arg, baseBlockPosFunction, model) {
        this.color = color
        this.map = function
    }

    constructor(
        arg: Properties,
        color: DyeColor,
        function: Map<DyeColor, Block>,
        model: ResourceLocation,
        width: Int,
        height: Int,
        length: Int
    ) : super(arg, model = model, width = width, height = height, length = length) {
        this.color = color
        this.map = function
    }

    constructor(
        color: DyeColor,
        function: Map<DyeColor, Block>,
        arg: Properties,
        model: ResourceLocation
    ) : super(arg, model = model) {
        this.color = color
        this.map = function
    }

    override fun useItemOn(
        stack: ItemStack,
        state: BlockState,
        level: Level,
        pos: BlockPos,
        player: Player,
        hand: InteractionHand,
        hitResult: BlockHitResult
    ): ItemInteractionResult {
        var state = state
        var pos: BlockPos = pos

        if (!level.isClientSide() && hand == InteractionHand.MAIN_HAND) {
            pos = getBaseBlockPos(pos, state)
            state = level.getBlockState(pos)
            val block = state.block

            return if (block is DyeableBlock<*, *> && !block.tryDyeColor(state, level, pos, player, hand, hitResult)) {
                block.serverUse(
                    stack,
                    state,
                    level as ServerLevel,
                    pos,
                    player as ServerPlayer,
                    hand,
                    hitResult
                )
            } else ItemInteractionResult.SUCCESS
        }
        return ItemInteractionResult.FAIL
    }

    fun getItemFromDyeColor(color: DyeColor): Item {
        return getBlockFromDyeColor(color).asItem()
    }

    fun getBlockFromDyeColor(color: DyeColor): Block = map[color]!!

    fun tryDyeColor(
        state: BlockState,
        world: Level,
        pos: BlockPos,
        player: Player,
        handIn: InteractionHand,
        hit: BlockHitResult
    ): Boolean {
        val heldItem = player.mainHandItem

        val isEmpty = !heldItem.isEmpty
        val isDye = heldItem.item is DyeItem

        if (isEmpty && isDye) {
            val dyeColor = (heldItem.item as DyeItem).dyeColor

            val baseState = world.getBlockState(pos)

            if (javaClass.isInstance(baseState.block)) {
                val baseBlock: DyeableBlock<*, *> = javaClass.cast(baseState.block)


                if (baseBlock.color != dyeColor) {
                    val base = getBaseBlockPos(pos, state)

                    if (!player.isCreative) heldItem.shrink(1)

                    val newBlock = getBlockFromDyeColor(dyeColor).instanceOrNull<DyeableBlock<*, *>>() ?: return false //TODO: This is a stop gap due to codec(). Yea its that expansive.

                    val defaultState = newBlock.defaultBlockState().applyIfTrue({ it.hasProperty(FACING) }, { it.setValue(FACING, baseState.getValue(FACING)) } )

                    val dir = state.getValue(FACING)

                    for (x in 0 until width + 1) {
                        for (y in 0 until height + 1) {
                            for (z in 0 until length + 1) {
                                val adjustX = adjustX(x)
                                val adjustZ = adjustX(x)

                                val blockPos = base.relative(dir.counterClockWise, adjustX).relative(Direction.UP, y)
                                    .relative(dir, adjustZ)

                                val currentState = world.getBlockState(blockPos)

                                val stateX = baseBlock.getWidthValue(currentState)
                                val stateY = baseBlock.getHeightValue(currentState)
                                val stateZ = baseBlock.getLengthValue(currentState)

                                world.setBlock(
                                    blockPos, newBlock.setSize(
                                        defaultState.setValue(
                                            WATERLOGGED, currentState.getValue(
                                                WATERLOGGED
                                            )
                                        ), stateX, stateY, stateZ
                                    ), 2, 0
                                )
                            }
                        }
                    }
                    return true
                }
            }
        }

        return false
    }

    override fun getCloneItemStack(level: LevelReader, pos: BlockPos, state: BlockState): ItemStack {
        return ItemStack(getItemFromDyeColor(color))

    }

    protected open fun serverUse(
        stack: ItemStack,
        state: BlockState,
        world: ServerLevel,
        pos: BlockPos,
        player: ServerPlayer,
        handIn: InteractionHand,
        hit: BlockHitResult
    ): ItemInteractionResult {
        return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION
    }

    override fun getDrops(state: BlockState, params: LootParams.Builder): List<ItemStack> {
        return if (getWidthValue(state) == 0 || getHeightValue(state) == 0 || getLengthValue(state) == 0) super.getDrops(
            state,
            params
        ) else emptyList()
    }

    override fun getVariant(): String? {
        return color.serializedName
    }

    companion object {
        fun <T: DyeableBlock<*, *>> simpleDyedCodec(supplier: (Properties, DyeColor, Map<DyeColor, Block>) -> T): MapCodec<T> = RecordCodecBuilder.mapCodec<T> { dyedCodecBuilder(it).apply(it, supplier::invoke) }

        fun <T: DyeableBlock<*, *>> dyedCodecBuilder(instance: RecordCodecBuilder.Instance<T>): Products.P3<RecordCodecBuilder.Mu<T>, Properties, DyeColor, MutableMap<DyeColor, Block>> {
            return instance.group(
                propertiesCodec(),
                DyeColor.CODEC.fieldOf("color").forGetter { it.color },
                Codec.unboundedMap(DyeColor.CODEC, BuiltInRegistries.BLOCK.byNameCodec()).fieldOf("map").forGetter { it.map }
            )
        }
    }
}
